### PR TITLE
[12.0][FIX] sale: Avoid performance/side effects when invoicing

### PR DESCRIPTION
**Pull request by pedrobaeza**
~Steps to reproduce the bug:~

~- Create two fiscal postions fp1 and fp2~
~- Create a customer C with fp1~
~- Create a SO for C and set fp2~
~- Create an invoice for the SO~

~BUG: The fiscal position of the SO was set to fp1.~

~Now the fiscal position of the SO is kept when creating an invoice from a SO (technically, is changed and restored back).~

~It was solved in c2f15d3, but due to the refactorization introduced in 389bd53, it was lost again due to merge slots.~

~This code is very ugly and under-performant (calling 2 times to compute taxes, an onchange that writes some unneeded data, and then write back again other data), but I have taken the occasion to at least improve a bit this part not writing again if the data hasn't changed and doing only one write in that case although both fields change.~

See https://github.com/odoo/odoo/pull/40156#issuecomment-555514751 for an explanation of why doing this.

cc @Tecnativa TT20585

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@simongoffin as you were the creator of the original patch at #32713, can you please review it?

@dbo-odoo you should be interested as well.

**Comment by pedrobaeza** [comment]
Would you accept the patch of the second commit instead for avoiding this and other side effects? If so, tell me and I squash commits and adapt commit message.

Look for example one of the side effects with current method on a community module extending this invoice generation: https://github.com/OCA/sale-workflow/pull/992

File: addons/sale/models/sale.py [hunk 1]
```diff
@@ -469,12 +469,15 @@ def _prepare_invoice(self):
             .default_get(['journal_id'])['journal_id'])
         if not journal_id:
             raise UserError(_('Please define an accounting sales journal for this company.'))
-        invoice_vals = {
+        vinvoice = self.env['account.invoice'].new({'partner_id': self.partner_invoice_id.id})
+        # Get partner extra fields
+        vinvoice._onchange_partner_id()
+        invoice_vals = vinvoice._convert_to_write(vinvoice._cache)
+        invoice_vals.update({
```
Comment (rousseldenis) [inline addons/sale/models/sale.py]: This code move makes sense as in a 'prepare' function!

**Comment by pedrobaeza** [comment]
@simongoffin can you please provide any feedback about this?

**Comment by simongoffin** [comment]
Hello @pedrobaeza I'm ok with your first commit  but for the second one can you detail it in an other PR?

**Comment by pedrobaeza** [comment]
@simongoffin but both are connected, as if we follow the approach of the second one, the first one is not needed. The question is: are you going to accept that approach? If not, then the first commit is restoring what you previously did, but I beg you to consider the second approach for avoiding more of these side effects.

**Comment by simongoffin** [comment]
@pedrobaeza  I'm not ok with the second approach because I don't get the business meaning of your side effects in Odoo standard and the code is a bit ugly. If you remove the second commit, I will merge it to restore the fix  c2f15d3

**Comment by pedrobaeza** [comment]
@simongoffin let me try to explain you again about the meaning of the second commit because I think it's important to fix it that way:

- On core, you have payment term (PT) and fiscal position (FP), that is put at partner level, SO level and invoice level.
- Partner info is used as base for both SO and invoice when you set the partner on them (via an onchange). But other values are set in the invoice onchange that are invoice specific, like invoice address one.
- When creating the invoice, the initial dictionary is created with the SO PT and FP.
- Then after being created, the rest of the values that we have talked about were not transferred, so in that process (inside `_finalize_invoices`), the onchange is called for getting them (and maybe as a later bugfix).
- But this also resets both PT and FP put on the SO, giving more priority to the partner ones.
- And thus you need to store valid PT and FP values before calling onchange, and restoring them after.

Simply on core, this means a waste of resources as you need at least an extra write after creating the invoice.

But not only that, it can create side effects when extending sales invoicing. I have already 2 documented side effects on OCA modules:

- On [`sale_order_type`](https://github.com/OCA/sale-workflow/tree/12.0/sale_order_type):
  - We add a new field the same way at partner, SO and invoice called "order type" (OT).
  - As we want to use the SO one on invoice, we have to overwrite the `_finalize_invoice` method, storing before calling super the current OT, and writing it back if changed, performing an extra write if so. PR for doing that proposed already in https://github.com/OCA/sale-workflow/tree/12.0/sale_order_type (but preferable to avoid it).

- On [`account_invoice_fiscal_position_update`](https://github.com/OCA/account-invoicing/tree/12.0/account_invoice_fiscal_position_update):
  - As on core, the fiscal position is potentially overwritten 2 times, we trigger the recomputation of taxes these 2 times without being needed, due to the way is done as explained before.

Thus, the ideal solution is to not delay the fetch of this extra data to that phase, but doing early when preparing the values dictionary. This way, we get the desired values, and overwriting the rest that are provided by the SO.

That can be done performing the onchange on a virtual record and getting the values after that from the cache. It might look ugly, but it's the method that is used on several places across Odoo. Some examples:

https://github.com/odoo/odoo/commit/47496b9fe5555ef23932fcfc3f2d6d6ee945daae#diff-f3d55d07b488d54ad49be4799d8f276cR387-R390
https://github.com/odoo/odoo/commit/4357afa94b908ae4b83d0ef91393973e0e3362c9#diff-2d656902930900b836412f086c924643R245-R247
https://github.com/odoo/odoo/blob/84cf3ceb449a8adb4d6a182b19f48395b8db636c/addons/point_of_sale/models/pos_order.py#L150-L159

I hope this clarifies why the change and the importance of doing this patch for not having to need any other later bugfix nor any penalty on extending the SO invoicing process.

**Comment by pedrobaeza** [comment]
@simongoffin did you have the chance to take a look to my last comment?

**Comment by pedrobaeza** [comment]
@simongoffin my weekly reminder for checking this serious one :)

**Comment by Yajo** [comment]
This has conflicts now.

**Comment by Yajo** [comment]
Conflicts fixed.

File: addons/sale/models/sale.py (hunk 1 above)
Comment (pedrobaeza) [reply addons/sale/models/sale.py]: It's already in a prepare function :wink:

**Comment by pedrobaeza** [comment]
@dbo-odoo as the bugfix team has already fixed the underlying bug via an OPW in #42151, can you take a look at this need improvement reading the rationale behind it?

Thanks and Merry Christmas!

**Comment by dbo-odoo** [comment]
@simongoffin 
In this case, I agree with @pedrobaeza - calling the onchange after the invoice is created is bound to cause issue and to force devs to save/reset their field around this method, which is not only bad performance-wise but really ugly from an inheritance/modularity PoV.

The solution proposed here seems quite alright - not the most readable one, I agree, but this is not some dark magic (it's using the ORM properly).

I would honestly have preferred that the dict of values generated by the `sale` module would be complete, but I see that `l10n_mx` and `account_intrastat` both add some overrides on the onchange which should still be called during invoice creation. So, making sure the dict is complete is not an option.

Here's what I suggest:
- @simongoffin if you really are against it, please say so and why
- @Yajo: please add comments (`# here we use a cached record to apply the onchange; we can then properly override the content of the record instead of calling the onchange once the record actually exists`) and a better commit message explaining why this change is needed (the current commit message does not match the content). Here's a suggestion (feel free to ignore it and to write your own as long as it explains *why* you're doing this change):
```git
[FIX] sale: apply onchange before creating an invoice

Up to now, the method `_finalize_invoices` explicitely called the
onchange method on the partner to ensure that all business fields
were set correctly. This is necessary due to other modules that may
extend this onchange to add specific fields required by their inner working.

Unfortunately, calling the onchange after the invoice has been created has
several drawbacks. It forces us to store existing values for these fields in a
temporary variable then re-assign it to the invoice immediately after the
onchange. This is not only bad performance wise (multiple writes instead of
a single one), but also from an inheritance point of view, since it makes it difficult
to introduce new fields impacted by the same onchange - you then also have to
override the finalization method to save/restore your fields as well.

This commit move this call to before the creation of the invoice, by using
a cached record on which to apply the onchange then converting this cached
record to a dictionnary; the invoice preparation method can then go on normally,
overriding any result of the onchange with the correct values.
```

**Comment by rousseldenis** [comment]
@simongoffin

**Comment by simongoffin** [comment]
robodoo r+

**Comment by pedrobaeza** [comment]
Thanks for considering it, @dbo-odoo! But I'm afraid it has landed already with the previous commit message without your suggestions (which are very good BTW, as this PR has walked a long way and PR message ). Should we amend it at least on forward-ports?

**Comment by simongoffin** [comment]
Hello @pedrobaeza 
Sorry I've just noticed your last comment now.

**Comment by pedrobaeza** [comment]
The important thing is that the fix has landed :) But indeed when doing git blame, it will be weird to see that comment.

**Pull request closed without merging**